controllers/volumereplicationgroup_controller.go

```diff
@@ -462,27 +468,132 @@ func (v *VRGInstance) createVolumeReplicationResources() bool {
-			requeue = true
+			if err = v.createVolumeReplicationForPVC(pvc); err == nil {
+				log.Error(err, "failed to upload PV metadata")
+
+				continue
+			}
 		}
+
+		log.Info("Requeuing due to failure in getting or creating VolumeReplication resource for PersistentVolumeClaim",
+			"errorValue", err)
+
+		// requeue on failure to ensure any PVC not having a corresponding VR CR
+		requeue = true
 	}
 
 	return requeue
 }
 
+// uploadPV checks if the VRG spec has been configured with an s3 endpoint,
+// validates the S3 endpoint, connects to it, gets the PV metadata of
+// the input PVC, creates a bucket in s3 store, upload's the PV metadata to
+// s3 store and downloads it for verification.  If an s3 endpoint is not
+// configured, then it assumes that VRG is running in a backup-less mode and
+// does not return an error, but logs a one-time warning.
+func (v *VRGInstance) uploadPV(pvc corev1.PersistentVolumeClaim) error {
+	// Create an s3 bucket name from the VRG's namespace and name, separating
+	// the two with a forward slash.  Note: VRG's namespace and name may also
+	// have dots or hyphens in their names.
+	vrgName := v.instance.Name
+	s3Bucket := v.instance.Namespace + "/" + vrgName
+
+	s3Endpoint := v.instance.Spec.S3Endpoint
+	if err := validateS3Endpoint(s3Endpoint, s3Bucket, v.log); err != nil {
+		return err
+	} else if s3Endpoint == "" {
+		// VRG is configured to run in a backup-less mode.  validateS3Endpoint()
+		// logs a warning message once about this backup-less mode; hence, no
+		// need to log any error here.
+		return nil
+	}
+
+	s3Conn, err := connectToS3Endpoint(v.ctx, v.reconciler,
+		s3Endpoint,
+		types.NamespacedName{ /* secretName */
+			Name:      v.instance.Spec.S3SecretName,
+			Namespace: v.instance.Namespace,
+		},
+		vrgName, /* debugTag */
+	)
+	if err != nil {
+		return err
+	}
+
+	pv := corev1.PersistentVolume{}
+	volumeName := pvc.Spec.VolumeName
+	pvObjectKey := client.ObjectKey{
+		Name: volumeName,
+	}
+
+	if err := v.reconciler.Get(v.ctx, pvObjectKey, &pv); err != nil {
+		return fmt.Errorf("failed to get PV metadata from k8s, %w", err)
+	}
+
+	// Create the s3Bucket, without caching or assuming its existence
+	if err := s3Conn.createBucket(s3Bucket); err != nil {
+		return fmt.Errorf("unable to create s3Bucket %s, %w",
+			s3Bucket, err)
+	}
+
+	if err := s3Conn.uploadPV(s3Bucket, pv.Name, pv); err != nil {
+		return err
+	}
+
+	if err := s3Conn.verifyPVUpload(s3Bucket, pv.Name, pv); err != nil {
+		return err
+	}
+
+	// Remove after Benamar uses this method in AVR
+	if _, err := s3Conn.downloadPVs(s3Bucket); err != nil {
+		return err
+	}
+
+	return err
+}
+
+// If the the s3 endpoint is not set, then the VRG has been configured to run
+// in a backup-less mode to simply control VR CRs alone without backing up
+// PV k8s metadata to an object store.
+func validateS3Endpoint(s3Endpoint, s3Bucket string, log logr.Logger) error {
```

Comment:
(future) This utility function `validateS3Endpoint` should validate that the endpoint is valid and no more.  An empty s3Endpoint is **invalid** in my mind.  This takes us back to my comment in line 520. If the `else if` clause is before line 518, then you don't need to worry about s3Endpoint emptiness and this function becomes even simpler.

Reply:
The reason it is coded in this admittedly peculiar way is to have to logic of one-time log.Warn message of running in backup-less mode inside the validateS3Endpoint method.  UploadPV is very close to 60 lines.